### PR TITLE
[SYNPY-1156] dateTime_dataframe_store_to_Synapse

### PR DESCRIPTION
**Pull request by linchiahuisage**
Change

- if the column type is DATE, change the date format to unix epoch and trailing the 3 digit of millisecond

### CODE REVIEW

File: synapseclient/table.py [hunk 1]
```diff
@@ -1756,6 +1756,15 @@ def from_data_frame(cls, schema, df, filepath=None, etag=None, quoteCharacter='"
 
             f = io.open(filepath, mode='w', encoding='utf-8', newline='')
 
+            import pandas as pd
```
Comment (thomasyu888) [inline synapseclient/table.py:1759]: This sort of makes pandas a required install right?  Should there be a check for if pandas is installed?

Comment (jkiang13) [reply synapseclient/table.py:1759]: For consistency we should use the same check although it's in the from_data_frame which is likely only going to be called from a context where pandas was already used.

I'm not sure this is the best approach though, Godfrey was working on this as he left and didn't have a chance to finish it. At very least if we modify the df for the purposes of serializing it to a csv file we should copy it first. The to_csv function of the data frame has a date_format param on it to which we might be able to use instead although it I'm not sure there is an easy way to preserve millisecond accuracy using that string format alone...

File: tests/integration/synapseclient/test_tables.py [hunk 2]
```diff
@@ -233,7 +233,10 @@ def test_tables_pandas(syn, project):
     df['string_'] = df['string_'].transform(str)
 
     # SYNPY-717
-    df['datetime64'] = df['datetime64'].apply(lambda x: pd.Timestamp(x).tz_localize('UTC'))
+    if os.name == 'nt':
+        df['datetime64'] = pd.to_datetime(df['datetime64'], utc=True)
+    else:
+        df['datetime64'] = pd.to_datetime(df['datetime64'], unit='ms', utc=True)
```
Comment (mattfazza) [inline tests/integration/synapseclient/test_tables.py:239]: I would add a comment that this is a check for Windows.

File: synapseclient/table.py (hunk 1 above)
Comment (mattfazza) [reply synapseclient/table.py:1759]: I'm late to this conversation, but what's the problem of requiring pandas?

**Review by mattfazza — COMMENTED** [review]
Left a couple of comments.  Everything looks fine though.

File: synapseclient/table.py (hunk 1 above)
Comment (thomasyu888) [reply synapseclient/table.py:1759]: @mattfazza The python package actually doesn't require the installation of pandas.  So the parts of the client where pandas is needed `def asDataFrame(...)` actually uses this: https://github.com/Sage-Bionetworks/synapsePythonClient/blob/develop/synapseclient/table.py#L1455

File: tests/integration/synapseclient/test_tables.py [hunk 3]
```diff
@@ -233,7 +233,10 @@ def test_tables_pandas(syn, project):
     df['string_'] = df['string_'].transform(str)
 
     # SYNPY-717
-    df['datetime64'] = df['datetime64'].apply(lambda x: pd.Timestamp(x).tz_localize('UTC'))
+    if os.name == 'nt':
```
Comment (thomasyu888) [inline tests/integration/synapseclient/test_tables.py:236]:
```suggestion
# This is a check for windows
    if os.name == 'nt':
```

File: synapseclient/table.py (hunk 1 above)
Comment (thomasyu888) [inline synapseclient/table.py:1759]:
```suggestion
            test_import_pandas()
```

File: tests/integration/synapseclient/test_tables.py [hunk 4]
```diff
@@ -233,7 +233,11 @@ def test_tables_pandas(syn, project):
     df['string_'] = df['string_'].transform(str)
 
     # SYNPY-717
-    df['datetime64'] = df['datetime64'].apply(lambda x: pd.Timestamp(x).tz_localize('UTC'))
+# This is a check for windows
+    if os.name == 'nt':
```
Comment (thomasyu888) [inline tests/integration/synapseclient/test_tables.py:237]:
```suggestion
    # This is a check for windows
    if os.name == 'nt':
```

File: synapseclient/table.py [hunk 5]
```diff
@@ -1756,6 +1756,15 @@ def from_data_frame(cls, schema, df, filepath=None, etag=None, quoteCharacter='"
 
             f = io.open(filepath, mode='w', encoding='utf-8', newline='')
 
+            test_import_pandas()
```
Comment (thomasyu888) [inline synapseclient/table.py:1759]:
```suggestion
            test_import_pandas()
            import pandas as pd
```

**Review by emmanmills — APPROVED** [review]

**Pull request merged**